### PR TITLE
Fixed clone function to clone innerBlocks.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Clone function did not clone the inner blocks so if we clone a block e.g: a column with inner blocks both columns would reference the same inner blocks so changes applied to one inner block would also apply to the other.
Now if inner blocks are not passed explicitly to clone, all the inner blocks are recursively cloned.
Props to @mcsf for discovering this bug.

## How Has This Been Tested?
Add a column block, add a paragraph inside it, clone the column block, change the text of the cloned block and verify it does not changes the content of the original paragraph.

## Screenshot of the bug
apr-04-2018 22-27-34

### CODE REVIEW

**Review by mcsf — APPROVED** [review]
Thanks for the quick fix :)

File: blocks/api/factory.js [hunk 1]
```diff
@@ -71,15 +71,17 @@ export function createBlock( name, blockAttributes = {}, innerBlocks = [] ) {
  *
  * @return {Object} A cloned block.
  */
-export function cloneBlock( block, mergeAttributes = {}, innerBlocks = block.innerBlocks ) {
+export function cloneBlock( block, mergeAttributes = {}, innerBlocks ) {
```
Comment (aduth) [inline blocks/api/factory.js]: Could we just keep the default value and just continue with the `Array#map` clone, avoiding the truthiness checking below?

File: blocks/api/factory.js [hunk 2]
```diff
@@ -71,15 +71,17 @@ export function createBlock( name, blockAttributes = {}, innerBlocks = [] ) {
...
 		...block,
 		uid: uuid(),
 		attributes: {
 			...block.attributes,
 			...mergeAttributes,
 		},
-		innerBlocks,
+		innerBlocks: innerBlocks ?
+			innerBlocks :
+			( block.innerBlocks || [] ).map( ( innerBlock ) => cloneBlock( innerBlock ) ),
```
Comment (aduth) [inline blocks/api/factory.js]: When will `innerBlocks` be falsey?

File: blocks/api/factory.js (hunk 1 above)
Comment (jorgefilipecosta) [reply blocks/api/factory.js]: Hi @aduth, if we do this change even when innerBlocks are explicitly passed, we will clone them. Clonning them will change their uid's, when I call cloneBlocks and I explicitly pass the innerBlocks array I expect that the uids I passed are respected that's why I changed this logic to not clone the blocks if they are explicitly passed.

[the remark this answers is not in the document]
Comment (jorgefilipecosta) [reply blocks/api/factory.js]: Improvement applied.

File: blocks/api/factory.js [hunk 3]
```diff
@@ -71,15 +71,17 @@ export function createBlock( name, blockAttributes = {}, innerBlocks = [] ) {
...
 	return {
 		...block,
 		uid: uuid(),
 		attributes: {
 			...block.attributes,
 			...mergeAttributes,
 		},
-		innerBlocks,
+		innerBlocks: innerBlocks ?
+			innerBlocks :
```
[the remark this answers is not in the document]
Comment (jorgefilipecosta) [reply blocks/api/factory.js]: Yes totally right, missed this logic simplification.

File: blocks/api/factory.js (hunk 2 above)
Comment (jorgefilipecosta) [reply blocks/api/factory.js]: Only if the block was wrongly changed/created, but yes we should not be "defending" against this case, I removed this check.

**Pull request merged**